### PR TITLE
API Response returns specialized error type

### PR DESCRIPTION
**Pull request by dadgar**
This PR adds a specialized error object that is returned by the Vault API client. The error type exposes the HTTP method, URL, HTTP status code as well as the individual errors returned by Vault.

Further the error type exposes helpers to extract gRPC status codes. These helpers make it easy to integrate Vault into systems using gRPC and status error messages. Happy to remove those if you don't think they will be useful to the broader community. Reason I added is that it will just be generic boilerplate everyone who does use gRPC status codes will have to do.

### CODE REVIEW

**Comment by jefferai** [comment]
I don't think the grpc codes belong here. The API isn't grpc. I also think different people will want different behavior for different purposes and I don't want to get into a situation of having to add toggles for these. (E.g. not everyone will necessarily want to consider a standby as ok.)

**Comment by mitchellh** [comment]
I agree with @jefferai. We can implement the gRPC code mapping in our packages downstream. I think richer errors are necessary though, we're having a really tough time in `[secret team]` discerning retryable vs. non-retryable and user vs. system errors.

File: api/response.go [hunk 1]
```diff
@@ -42,36 +42,78 @@ func (r *Response) Error() error {
...
+		"Error making API request.\n\n"+
+			"URL: %s %s\n"+
+			"Code: %d. %s:\n\n",
+		r.HTTPMethod, r.URL, r.StatusCode, errString))
+
+	if r.RawError && len(r.Errors) == 1 {
+		errBody.WriteString(r.Errors[0])
+	} else {
+		for _, err := range r.Errors {
+			errBody.WriteString(fmt.Sprintf("* %s", err))
```
Comment (kalafut) [inline api/response.go]: Should these be newline delimited?

File: api/response.go [hunk 2]
```diff
@@ -42,36 +42,78 @@ func (r *Response) Error() error {
...
-			"Code: %d. Errors:\n\n",
-		r.Request.Method, r.Request.URL.String(),
-		r.StatusCode))
-	for _, err := range resp.Errors {
-		errBody.WriteString(fmt.Sprintf("* %s", err))
-	}
-
-	return fmt.Errorf(errBody.String())
+	// Store the decoded errors
+	respErr.Errors = resp.Errors
```
Comment (kalafut) [inline api/response.go]: This will overwrite the JSON decode error special handling.

File: api/response.go (hunk 1 above)
Comment (dadgar) [reply api/response.go]: I had a similar thought but I just ported the formatting from the old code. I am assuming that the Vault errors are newline delimited. Happy to detect and ensure that there is a new line if you'd like?

Comment (kalafut) [reply api/response.go]: Oh ok, I was thinking about the go-multierror handling. Keeping the current approach is fine.

**Pull request merged**